articles/qos_configurability.md

```diff
@@ -0,0 +1,399 @@
+auto policy_x_value = node->declare_parameter(
+    "<name of the parameter>", qos_provided_in_code.<getter_for_policy_x>(), policy_x_description);
+// Repeat for all policies
+rclcpp::QoS final_qos();
+// construct `final_qos` from the parameter values read above
+if (!user_provided_callback(final_qos) {
+    // qos profile rejected, throw error here
+}
+```
+
+In that way, we avoid installing a parameter callback that will be triggered when any parameter is modified.
+
+Note: this seems to be what you generally want in the case of read-only parameters.
+Note 2: We could also have a `declare_parameters_atomically` method and use:
+  ```cpp
+  auto callback_handle = node->add_on_set_parameters_callback(wrapper_for_qos_callback(user_provided_callback));
+  node->declare_parameters_atomically(...);  // names and defaults for all parameters given here
+  // callback_handle goes out of scope, the callback is auto-removed.
+  ```
+  `declare_parameters_atomically` would be needed, if not you cannot have conditions like `history == keep_all or history_depth > 10`.
+
+### Parameters URIs
+
+Parameters URIs currently look like:
+
+```
+rosparam://my/full/namespace/node_name/param_group.nested_param_group.param_name
+```
+
+For the following QoS policy:
+- Full node name: `/asd/bsd/my_node`
+- Entity: `publisher`
+- Topic name: `/foo/bar/topic_name`
+- Policy: `reliability`
+
+The URI of the automatically generated parameter would be:
+
+```
+rosparam://asd/bsd/my_node/qos_profiles.publisher./foo/bar/topic_name.reliability
+```
+
+The `/` in one of the subgroups looks odd, but it's currently accepted.
+
+If in the future the addressing changes described in [#241](https://github.com/ros2/design/pull/241) are implemented, the URI would be:
+
+```
+rosparam://asd.bsd.my_node/qos_profiles/publisher/foo/bar/topic_name/reliability
+```
+
+but in both cases the previous parameters file example wouldn't change.
+
+## Caveats
+
+### Creating more than one publisher/subscription in the same topic with different QoS
+
+If only the topic name is used for the parameter name, it would be impossible to create two publishers with different QoS in the same topic.
+
+That could be solved by adding an optional extra identifier:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  QosOverridingOptions {
+    [] (const QoSProfile & qos) -> bool {
+        return qos.is_reliable();  // only accept reliable communication
+    },
+    "id1"
+  });
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  QosOverridingOptions { 
+    [] (const QoSProfile & qos) -> bool {
+      return !qos.is_reliable();  // only accept best effort communication
+    },
+    "id2"
+  });
+```
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:
+      publisher_id1:  # {entity_type}_{id}
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+            history_depth: 100
+             history: keep_last
```

Comment:
The indentation looks really weird here, maybe all of these three are supposed to be aligned?

Reply:
Fixed in https://github.com/ros2/design/pull/300/commits/0cee04571d97a0f8b746bbe0388e7a3aae284c82.